langkit/railroad_diagrams.py

```diff
@@ -0,0 +1,98 @@
+from os import path as P
+import pathlib
+from typing import Union
+
+from langkit.parsers import (
+    Defer, List, Opt, Or, Parser, _Extract, _Row, _Token, _Transform
+)
+
+
+def emit_railroad_diagram(parser: Parser) -> None:
+    """
+    Pass to emit railroad diagrams using the railroad_diagrams library.
+
+    Railroads will be emitted in $BUILD/doc/svg.
+    """
+    try:
+        from railroad import (
+            Choice, Diagram, DiagramItem, End, OneOrMore, Optional,
+            Sequence, Skip, Start, ZeroOrMore
+        )
+    except ImportError:
+        return
+
+    from langkit.compile_context import get_context
+
+    def recurse(p: Parser) -> Union[DiagramItem, str, None]:
+
+        # Transform parsers are just ignored
+        if isinstance(p, _Transform):
+            return recurse(p.children[0])
+
+        elif isinstance(p, Opt):
+            # Opt parsers are straightforwardly wrapped into an Optional
+            return Optional(recurse(p.parser))
+
+        elif isinstance(p, _Extract):
+            # Extract is ignored
+            return recurse(p.parser)
+
+        # For list parsers, we create a sequence with the right separator
+        # and sub-parser.
+        elif isinstance(p, List):
+
+            sep = recurse(p.sep) if p.sep else None
+            child = recurse(p.parser)
+            if p.empty_valid:
+                return ZeroOrMore(child, repeat=sep)
+            else:
+                return OneOrMore(child, repeat=sep)
+
+        # For defers, we just return the rule name
+        elif isinstance(p, Defer):
+            return p.rule_name
+
+        # For tokens, we return either the quoted original string, or the DSL
+        # name.
+        elif isinstance(p, _Token):
+            if p._original_string:
+                return repr(p._original_string)
+            else:
+                return p.val.dsl_name
+
+        children = []
+
+        for c in p.children:
+            res = recurse(c)
+            if res is not None:
+                children.append(res)
+
+        if isinstance(p, Or):
+            if len(children) == 0:
+                return None
+
+            children = sorted(children, key=lambda c: isinstance(c, Skip))
+            return Choice(0, *children)
+
+        elif isinstance(p, _Row):
+            if len(children) == 0:
+                return Skip()
+
+            return Sequence(*children)
+
+        else:
+            return None
+
+    d = Diagram(
+        # Explicit start point with the parser's name as label
+        Start("simple", label=parser.name.lower()),
+        *[c for c in [recurse(parser)] if c is not None],
+        End("simple")
+    )
+
+    # Output the diagram to svg in $BUILD/doc/svg/$RULENAME.svg
+    out_dir = P.join(get_context().emitter.lib_root, "doc", "svg")
```

Comment:
Please add a subdirectory that corresponds to the generated library (using `emitter.lib_name_low`)

Also, since `pathlib.Path` is involved, this block could be greatly simplified:

```python
out_dir = pathlib.Path(…, "doc", "svg")
out_dir.mkdir(parents=True, exist_ok=True)
with (out_dir / …).open("w"):
    …
```

Reply:
> Please add a subdirectory that corresponds to the generated library

Why ? I don't understand, it was my impression that we were *not* trying to mimic install trees anymore in `build/` ?

Reply:
It’s true we don’t try anymore for anything that’s under GPRbuild/GPRinstall’s control (sources, build artifacts), but the install procedure is manual for everything else. Given that we just merge `share` into the prefix at install type (in our manage scripts), `share` must already have the correct layout. The other option is to generate the SVG outside of this directory and amend the `do_install` function install it in an appropriate place.

Reply:
Ok but to be honest, for the moment those are just build artifacts, and meant to be then leveraged in doc generation, which isn't even part of the build tree. I don't think we want to ship those as is. WDYT ?

Reply:
Everything in `$BUILD/share` is installed, so if this is just a build artifact, maybe generate it in `$BUILD/obj`?

Reply:
Yeah but it's more like source, ie. I don't want it to disappear on gprclean, since it's generated by `manage.py`

Reply:
gprclean does not removes all files in object directories, and `$BUILD/obj/` isn’t the actual object directory anyway (it’s `$BUILD/obj/$BUILD_MODE/` anyway, so there should be no problem. If it’s a matter of principle, then just generate the SVG directly in `$BUILD/`?

Reply:
But it is in `$BUILD` ... (in `$BUILD/doc` but not in `share` anyway, so I still don't really see the problem)

Reply:
Ah, my bad then… there is a `doc` dir in `share`, so I suppose I confused both. What about `$BUILD/parser-railroads/$RULE.svg` to avoid the confusion?

Reply:
Ok